p2p/log.go

```diff
@@ -9,100 +14,131 @@ const (
 	MtDatasetLog = MsgType("dataset_log")
 )
 
+// DatasetLogRequestBody encapsulates options for requesting dataset history
+type DatasetLogRequestBody struct {
+	Ref    repo.DatasetRef
+	Limit  int
+	Offset int
+}
+
+// DatasetLogResponseBody encapsulates option for responding to a dataset history request
+type DatasetLogResponseBody struct {
+	History []repo.DatasetRef
+	Err     error
+}
+
 // RequestDatasetLog gets the log information of Peer's dataset
 func (n *QriNode) RequestDatasetLog(ref repo.DatasetRef, limit, offset int) ([]repo.DatasetRef, error) {
-	// id, err := n.Repo.Peers().IPFSPeerID(ref.Peername)
-	// if err != nil {
-	// 	return nil, fmt.Errorf("error getting peer IPFS id: %s", err.Error())
-	// }
-	// res, err := n.SendMessage(id, &Message{
-	// 	Type:    MtDatasetLog,
-	// 	Phase:   MpRequest,
-	// 	Payload: ref,
-	// })
-	// if err != nil {
-	// 	log.Debugf("send dataset log message error: %s", err.Error())
-	// 	return nil, err
-	// }
-
-	// data, err := json.Marshal(res.Payload)
-	// if err != nil {
-	// 	log.Debug(err.Error())
-	// 	return nil, err
-	// }
-
-	resref := []repo.DatasetRef{}
-	// err = json.Unmarshal(data, &resref)
-	// if len(resref) == 0 && err != nil {
-	// 	err = fmt.Errorf("no log found")
-	// }
-
-	return resref, nil
+
+	// get a list of peers to whom we will send the request
+	pids := n.ClosestConnectedPeers(ref.ProfileID, 15)
+	if len(pids) == 0 {
+		return nil, fmt.Errorf("no connected peers")
+	}
+
+	// create a channel on which to send the requests and receive the responses
+	replies := make(chan Message)
+
+	// create body of message:
+	body := DatasetLogRequestBody{
+		Ref:    ref,
+		Limit:  limit,
+		Offset: offset,
+	}
+
+	// create the request
+	req, err := NewJSONBodyMessage(n.ID, MtDatasetLog, body)
+	req = req.WithHeaders("phase", "request")
+	if err != nil {
+		log.Debug(err.Error())
+		return nil, err
+	}
+
+	// iterate through the peer ids
+	for _, pid := range pids {
+
+		if err = n.SendMessage(req, replies, pid); err != nil {
+			log.Debugf("%s err: %s", pid, err.Error())
+			continue
+		}
+
+		res := <-replies
+		dlrb := DatasetLogResponseBody{}
+		if err = json.Unmarshal(res.Body, &dlrb); err == nil {
+			if dlrb.Err != nil {
```

Comment:
Here I think a comment would be helpful, as I'm a little lost on the semantics. If a peer doesn't own the dataset, it'll return with an empty history list (len == 0), but not an error, right? And we expect only a single peer to ever return a non-empty history list, which is why we return immediately upon finding one, correct?